src/lib/CLValue2/Map.ts

```diff
@@ -0,0 +1,103 @@
+import { CLType, CLValue } from './Abstract';
+
+export class MapValueType<K extends CLType, V extends CLType> extends CLType {
+  innerKey: K;
+  innerValue: V;
+
+  constructor(keyType: K, valueType: V) {
+    super();
+    this.innerKey = keyType;
+    this.innerValue = valueType;
+  }
+
+  toString(): string {
+    return `Map (${this.innerKey.toString()}: ${this.innerValue.toString()})`;
+  }
+}
+
+export interface MapEntryType {
+  key: CLType;
+  value: CLType;
+}
+
+export interface MapEntryValue {
+  key: CLType;
+  value: CLType;
+}
+
+const isValueConstructor = (
+  v: Array<[CLValue, CLValue]> | [CLType, CLType]
+): v is Array<[CLValue, CLValue]> => {
+  return (
+    Array.isArray(v) &&
+    Array.isArray(v[0]) &&
+    v[0].length === 2 &&
+    !!v[0][0].clType &&
+    !!v[0][1].clType
+  );
+};
+
+export class MapValue<K extends CLValue, V extends CLValue> extends CLValue {
+  v: Array<[K, V]>;
```

Comment:
I think it would be better to use [Map](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Global_Objects/Map) here and return Map in `value()`